src/main/java/gregtech/api/recipes/GTRecipeInputCache.java

```diff
@@ -0,0 +1,63 @@
+package gregtech.api.recipes;
+
+import gregtech.api.GTValues;
+import gregtech.api.recipes.ingredients.GTRecipeInput;
+import gregtech.api.util.GTLog;
+import gregtech.common.ConfigHolder;
+import it.unimi.dsi.fastutil.objects.ObjectOpenHashSet;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+public class GTRecipeInputCache {
+
+    private static final int EXPECTED_CACHE_SIZE = 16384;
+    private static ObjectOpenHashSet<GTRecipeInput> INSTANCES;
+
+    public static boolean isCacheEnabled() {
+        return INSTANCES != null;
+    }
+
+    public static void enableCache() {
+        if (INSTANCES == null) {
+            INSTANCES = new ObjectOpenHashSet<>(EXPECTED_CACHE_SIZE, 1);
+            if (ConfigHolder.misc.debug || GTValues.isDeobfEnvironment())
+                GTLog.logger.info("GTRecipeInput cache enabled");
+        }
+    }
+
+    public static void disableCache() {
+        if (INSTANCES != null) {
+            if (ConfigHolder.misc.debug || GTValues.isDeobfEnvironment())
+                GTLog.logger.info("GTRecipeInput cache disabled; releasing {} unique instances", INSTANCES.size());
+            INSTANCES = null;
+        }
+    }
+
+    public static GTRecipeInput deduplicate(GTRecipeInput recipeInput) {
+        if (INSTANCES == null || recipeInput.isCached()) {
+            return recipeInput;
+        }
+        GTRecipeInput cached = INSTANCES.addOrGet(recipeInput);
+        if (cached == recipeInput) {
```

Comment:
Should this use `==` or `a.equals(b)` here? If `==` is intended, a comment explaining why would be appreciated for future reference.

Reply:
The refeq is correct here. It checks if the `recipeInput` is put into cache, for subsequent marking of `cached` flag.

I'll insert some javadocs and comment to explain the code better.